include/nbl/video/IDeviceMemoryBacked.h

```diff
@@ -1,44 +1,75 @@
-// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
 // This file is part of the "Nabla Engine".
 // For conditions of distribution and use, see copyright notice in nabla.h
-
-#ifndef __NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED__
-#define __NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED__
+#ifndef _NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED_
+#define _NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED_
 
 #include "IDeviceMemoryAllocation.h"
 #include <algorithm>
 
-#define VK_NO_PROTOTYPES
-#include <vulkan/vulkan.h>
-
 namespace nbl::video
 {
 
-//! Interface from which resources backed by IDeviceMemoryAllocation, such as ITexture and IGPUBuffer, inherit from
+//! Interface from which resources backed by IDeviceMemoryAllocation inherit from
 class IDeviceMemoryBacked : public virtual core::IReferenceCounted
 {
     public:
+        //! If you bound an "exotic" memory object to the resource, you might require "special" cleanups in the destructor
+        struct ICleanup
```

Comment:
Is the swapchain going to inherit from this object?

Reply:
no swapchain is gonna make a little "functor" like this
```cpp
class ISwapchain
{
   protected:
      struct CImageReference : struct IDeviceMemoryBacked::ICleanup
      {
         ~CImageReference()
         {
            m_swapchain->setImageDead(m_imageIndex); // cant remember what this function is called XD
         }
         
         core::smart_refctd_ptr<ISwapchain> m_swapchain;
         uint8_t m_imageIndex;
      };
      friend struct CImageReference; // instead of making the ICPUImage a friend
};
```
in order to now have the extra members in every `IGPUImage` that @deprilula28 introduced